### ISSUE TITLE
Unbounded memory growth for task tracking.

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/async-service#43]
## What was wrong?

Currently, both the `trio` and `asyncio` service appear to have unbounded memory growth when running background tasks.

## How can it be fixed?

We need logic that for a given task, will wait until all of it's children have completed and been cleaned up, and then remove the task from the DAG.

**Comment by pipermerriam** [comment]
Some notes:

### Race condition with the `trio` service.

In `asyncio` we use `ensure_future` to schedule tasks which gives us a reference to the task from within the context of `run_task`.

In `trio` we are not able to get a reference to the task until we are inside of it, so in the `TrioManager`, it is possible for a task to get scheduled but not actually be in the task DAG at the time of cancellation because it hasn't actually started yet.  Everything still behaves the same because trio is smart and handles the scoping of cancellations for us.

I believe that fixing this discrepancy is going to be a necessary fix because of the scenario where task `A` schedules child task `B` and then immediately exits.  `A` will see itself as being complete and if `B` has not started, it won't register as having any dependencies and will clean itself up **before** `B` starts.  In this scenario `A` exits before it is supposed to since it should wait for `B` to be complete, and `B` ends up being unable to place itself as a dependency of `A` since it would have already been cleaned up.

To fix this I think we'll need to inject some sort of pending reference to the new task that is only resolved once the task starts allowing `A` to know that it has a *pending* child that it needs to wait on.

### Locks

We definitely need some locking/synchronization primatives put in place around the DAG tracking.  See #44 

During cancellation we need to know that no new tasks will be scheduled once the cleanup process starts.  This isn't trivial since task scheduling is synchronous and the standard synchronization primatives are asynchronous...

**Comment by pipermerriam** [comment]
# Trio cancel scopes

In the `TrioManager` we use an individual cancel scope per task and a `trio.Event` to signal the task is *done*.

I think that some care will need to be taken in the trio implementation to ensure that the code which monitors child tasks and waits for them to be cleaned up is properly scoped with respect to the cancel scope.
